File: auditHub/src/main/java/com/thy/audithub/service/DashboardService.java
package com.thy.audithub.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.thy.audithub.dto.DashboardEntry;
import com.thy.audithub.dto.IssueRowDto;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
@Service
@RequiredArgsConstructor
public class DashboardService {

    private final ObjectMapper objectMapper;

    // key = mudurluk + "|" + yearMonth  (örn. "Açık Sistem Çöz. Md.|2026-04")
    private final Map<String, DashboardEntry> store = new ConcurrentHashMap<>();

    private static final String DATA_FILE = "data/dashboard.json";

    @PostConstruct
    public void load() {
        File file = new File(DATA_FILE);
        if (!file.exists()) return;
        try {
            List<DashboardEntry> entries = objectMapper.readValue(
                    file, new TypeReference<List<DashboardEntry>>() {});
            entries.forEach(e -> store.put(key(e.getMudurluk(), e.getYearMonth()), e));
            log.info("Dashboard: {} kayıt yüklendi", entries.size());
        } catch (IOException e) {
            log.warn("Dashboard verisi yüklenemedi: {}", e.getMessage());
        }
    }

    public void saveOrUpdate(String mudurluk, String yearMonth, List<IssueRowDto> rows) {
        DashboardEntry entry = new DashboardEntry();
        entry.setMudurluk(mudurluk);
        entry.setYearMonth(yearMonth);
        entry.setTotalTasks(rows.size());

        int fU = 0, fUD = 0, fH = 0;
        int iU = 0, iUD = 0, iH = 0;
        int bta = 0;

        for (IssueRowDto row : rows) {
            switch (safe(row.getAnalizFormatUygunluk())) {
                case "Uygun"        -> fU++;
                case "Uygun Değil"  -> fUD++;
                case "Hesaplanamaz" -> fH++;
                default -> { /* boş bırakıldı */ }
            }
            switch (safe(row.getAnalizIcerikOlgunluk())) {
                case "Uygun"        -> iU++;
                case "Uygun Değil"  -> iUD++;
                case "Hesaplanamaz" -> iH++;
                default -> { /* boş bırakıldı */ }
            }
            bta += toInt(row.getBtaTest()) + toInt(row.getBtaDev());
        }

        entry.setFormatUygun(fU);
        entry.setFormatUygunDegil(fUD);
        entry.setFormatHesap(fH);
        entry.setIcerikUygun(iU);
        entry.setIcerikUygunDegil(iUD);
        entry.setIcerikHesap(iH);
        entry.setBtaCount(bta);
        entry.setLastUpdated(LocalDateTime.now()
                .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm")));

        store.put(key(mudurluk, yearMonth), entry);
        persist();
        log.info("Dashboard güncellendi: {} / {}", mudurluk, yearMonth);
    }

    public List<DashboardEntry> getAll() {
        return new ArrayList<>(store.values());
    }

    public void updateCell(String mudurluk, String yearMonth, String field, String value) {
        DashboardEntry entry = store.getOrDefault(key(mudurluk, yearMonth), new DashboardEntry());
        entry.setMudurluk(mudurluk);
        entry.setYearMonth(yearMonth);

        try {
            int intValue = Integer.parseInt(value);
            switch (field) {
                case "totalTasks" -> entry.setTotalTasks(intValue);
                case "formatUygun" -> entry.setFormatUygun(intValue);
                case "formatUygunDegil" -> entry.setFormatUygunDegil(intValue);
                case "formatHesap" -> entry.setFormatHesap(intValue);
                case "icerikUygun" -> entry.setIcerikUygun(intValue);
                case "icerikUygunDegil" -> entry.setIcerikUygunDegil(intValue);
                case "icerikHesap" -> entry.setIcerikHesap(intValue);
                case "btaCount" -> entry.setBtaCount(intValue);
                default -> throw new IllegalArgumentException("Bilinmeyen alan: " + field);
            }
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Sayı dönüşümü başarısız: " + value);
        }

        entry.setLastUpdated(LocalDateTime.now()
                .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm")));
        store.put(key(mudurluk, yearMonth), entry);
        persist();
        log.info("Dashboard hücresi güncellendi: {}/{} - {} = {}", mudurluk, yearMonth, field, value);
    }

    // -----------------------------------------------------------------------

    private String key(String m, String ym) {
        return m + "|" + ym;
    }

    private String safe(String s) {
        return s == null ? "" : s;
    }

    private int toInt(String s) {
        try { return Integer.parseInt(s == null ? "0" : s.trim()); }
        catch (NumberFormatException ignored) { return 0; }
    }

    private void persist() {
        try {
            File dir = new File("data");
            if (!dir.exists()) dir.mkdirs();
            objectMapper.writerWithDefaultPrettyPrinter()
                    .writeValue(new File(DATA_FILE), new ArrayList<>(store.values()));
        } catch (IOException e) {
            log.error("Dashboard kaydedilemedi: {}", e.getMessage());
        }
    }
}

File: Services/DashboardService.cs
using System.Collections.Concurrent;
using System.Text.Json;
using AuditHub.Models;

namespace AuditHub.Services;

public class DashboardService(ILogger<DashboardService> logger)
{
    private readonly ConcurrentDictionary<string, DashboardEntry> _store = new();
    private const string DataFile = "data/dashboard.json";
    private readonly JsonSerializerOptions _jsonOpts = new() { WriteIndented = true, PropertyNameCaseInsensitive = true };

    public DashboardService(ILogger<DashboardService> logger, IConfiguration _) : this(logger)
    {
        Load();
    }

    private void Load()
    {
        if (!File.Exists(DataFile)) return;
        try
        {
            var json = File.ReadAllText(DataFile);
            var entries = JsonSerializer.Deserialize<List<DashboardEntry>>(json, _jsonOpts) ?? [];
            foreach (var e in entries)
                _store[Key(e.Mudurluk, e.YearMonth)] = e;
            logger.LogInformation("Dashboard: {Count} kayıt yüklendi", entries.Count);
        }
        catch (Exception ex)
        {
            logger.LogWarning("Dashboard verisi yüklenemedi: {Msg}", ex.Message);
        }
    }

    public void SaveOrUpdate(string mudurluk, string yearMonth, List<IssueRowDto> rows)
    {
        int fU = 0, fUD = 0, fH = 0, iU = 0, iUD = 0, iH = 0, bta = 0;

        foreach (var row in rows)
        {
            switch (row.AnalizFormatUygunluk)
            {
                case "Uygun":        fU++;  break;
                case "Uygun Değil":  fUD++; break;
                case "Hesaplanamaz": fH++;  break;
            }
            switch (row.AnalizIcerikOlgunluk)
            {
                case "Uygun":        iU++;  break;
                case "Uygun Değil":  iUD++; break;
                case "Hesaplanamaz": iH++;  break;
            }
            bta += ToInt(row.BtaTest) + ToInt(row.BtaDev);
        }

        var entry = new DashboardEntry
        {
            Mudurluk = mudurluk,
            YearMonth = yearMonth,
            TotalTasks = rows.Count,
            FormatUygun = fU, FormatUygunDegil = fUD, FormatHesap = fH,
            IcerikUygun = iU, IcerikUygunDegil = iUD, IcerikHesap = iH,
            BtaCount = bta,
            LastUpdated = DateTime.Now.ToString("dd.MM.yyyy HH:mm")
        };

        _store[Key(mudurluk, yearMonth)] = entry;
        Persist();
        logger.LogInformation("Dashboard güncellendi: {M}/{Y}", mudurluk, yearMonth);
    }

    public List<DashboardEntry> GetAll() => [.. _store.Values];

    public void UpdateCell(string mudurluk, string yearMonth, string field, string value)
    {
        if (!int.TryParse(value, out var intValue))
            throw new ArgumentException("Sayı dönüşümü başarısız: " + value);

        var entry = _store.GetOrAdd(Key(mudurluk, yearMonth), _ => new DashboardEntry
        {
            Mudurluk = mudurluk,
            YearMonth = yearMonth
        });

        switch (field)
        {
            case "totalTasks":       entry.TotalTasks = intValue;       break;
            case "formatUygun":      entry.FormatUygun = intValue;      break;
            case "formatUygunDegil": entry.FormatUygunDegil = intValue; break;
            case "formatHesap":      entry.FormatHesap = intValue;      break;
            case "icerikUygun":      entry.IcerikUygun = intValue;      break;
            case "icerikUygunDegil": entry.IcerikUygunDegil = intValue; break;
            case "icerikHesap":      entry.IcerikHesap = intValue;      break;
            case "btaCount":         entry.BtaCount = intValue;         break;
            default: throw new ArgumentException("Bilinmeyen alan: " + field);
        }

        entry.LastUpdated = DateTime.Now.ToString("dd.MM.yyyy HH:mm");
        Persist();
    }

    private static string Key(string m, string ym) => $"{m}|{ym}";

    private static int ToInt(string? s)
    {
        return int.TryParse(s?.Trim(), out var v) ? v : 0;
    }

    private void Persist()
    {
        try
        {
            Directory.CreateDirectory("data");
            File.WriteAllText(DataFile, JsonSerializer.Serialize(_store.Values.ToList(), _jsonOpts));
        }
        catch (Exception ex)
        {
            logger.LogError("Dashboard kaydedilemedi: {Msg}", ex.Message);
        }
    }
}
